pkg/river/token/builder/blocks_nesting_test.go

```diff
@@ -0,0 +1,693 @@
+			// is a struct, so it will be initialized to zero value as we create a new outer block.
+			//TODO: should we document this behavior better?
+			name:  "no value set",
+			in:    OutNoDefInStrBlkOptWithDef{},
+			river: "",
+		},
+		{
+			name: "different value set",
+			in: OutNoDefInStrBlkOptWithDef{
+				Inner: AttrWithDefault{Number: otherDefaultNumber},
+			},
+			river: `
+			inner {
+				number = 321
+			}
+		`,
+		},
+		{
+			// NOTE: this is correct, because inner block's defaults will be applied to empty body `{ }`
+			name: "default value set",
+			in: OutNoDefInStrBlkOptWithDef{
+				Inner: AttrWithDefault{Number: defaultNumber},
+			},
+			river: `inner { }`,
+		},
+	}
+}
+
+// OutDiffDefInStrBlkOptDiffDef - outer with different default, inner struct block, optional with a different default value
+type OutDiffDefInStrBlkOptDiffDef struct {
+	Inner AttrWithDefault `river:"inner,block,optional"`
+}
+
+func (o *OutDiffDefInStrBlkOptDiffDef) SetToDefault() {
+	*o = OutDiffDefInStrBlkOptDiffDef{Inner: AttrWithDefault{Number: otherDefaultNumber}}
+}
+
+func (o *OutDiffDefInStrBlkOptDiffDef) testCases() []testCase {
+	return []testCase{
+		{
+			name: "no value set",
+			in:   OutDiffDefInStrBlkOptDiffDef{},
+			river: `
+		inner {
+			number = 0
+		}`,
+		},
+		{
+			name: "different value set",
+			in: OutDiffDefInStrBlkOptDiffDef{
+				Inner: AttrWithDefault{Number: 42},
+			},
+			river: `
+			inner {
+				number = 42
+			}
+		`,
+		},
+		{
+			// NOTE: again, when we provide empty body `{ }`, the inner block's defaults will be applied
+			name: "inner default value set",
+			in: OutDiffDefInStrBlkOptDiffDef{
+				Inner: AttrWithDefault{Number: defaultNumber},
+			},
+			river: `inner { }`,
+		},
+		{
+			// NOTE: when we don't provide anything, the outer block's defaults will be applied
+			name: "outer default value set",
+			in: OutDiffDefInStrBlkOptDiffDef{
+				Inner: AttrWithDefault{Number: otherDefaultNumber},
+			},
+			river: ``,
+		},
+	}
+}
+
+// ========== tests with inner pointer to struct ==========
+
+// OutZeroDefInPtrBlkOptWithDef - outer with zero value default, inner struct pointer, optional with a default value
+type OutZeroDefInPtrBlkOptWithDef struct {
+	Inner *AttrWithDefault `river:"inner,block,optional"`
+}
+
+func (o *OutZeroDefInPtrBlkOptWithDef) SetToDefault() {
+	*o = OutZeroDefInPtrBlkOptWithDef{Inner: &AttrWithDefault{}}
+}
+
+func (o *OutZeroDefInPtrBlkOptWithDef) testCases() []testCase {
+	return []testCase{
+		//TODO: invariant violated.
+		// The test case is: outer block has zero value default and a pointer to inner block. The inner block has a
+		// default value. So the outer block's default set the inner to nil.
+		// Seems impossible to encode this case in River, because we would need to somehow explicitly set the inner block
+		// to nil? How can we do that?
+
+		//{
+		//	name: "nil",
+		//	in: OutZeroDefInPtrBlkOptWithDef{
+		//		Inner: nil,
+		//	},
+		//	river: ``,
+		//},
```

Comment:
This is another outstanding issue. I'm not sure how we can fix it. 
We could return an error in decode when a block field with defaults is set to nil.
Or we could disallow pointers to have defaults - which would be a bigger change.